### PR TITLE
feat: MacOS support with the nil backend

### PR DESCRIPTION
**Pull request by haraldh**
Fixes: #1790 
```
❯ uname -a
Darwin MacBook-Pro-3.fritz.box 21.4.0 Darwin Kernel Version 21.4.0: Fri Mar 18 00:45:05 PDT 2022; root:xnu-8020.101.4~15/RELEASE_X86_64 x86_64

❯ cargo test --no-default-features --features=backend-nil --test wasmldr_tests
    Finished test [unoptimized + debuginfo] target(s) in 0.38s
     Running tests/wasmldr_tests.rs (target/debug/deps/wasmldr_tests-6c8348af9ae6e231)

running 9 tests
test memspike ... ok
test check_tcp ... ok
test return_1 ... ok
test hello_wasi_snapshot1 ... ok
test memory_stress_test ... ok
test wasi_snapshot1 ... ok
Error: default export in '' is not a function
test no_export ... ok
test echo ... ok
test zerooneone ... ok

test result: ok. 9 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 24.14s

```

Signed-off-by: Harald Hoyer <harald@profian.com>

### CODE REVIEW

**Comment by platten** [comment]
When trying to run on an M1 mac:
```bash
❯ cargo test --no-default-features --features=backend-nil --test wasmldr_tests
   Compiling memchr v2.4.1
   Compiling parking_lot_core v0.8.5
   Compiling lock_api v0.4.7
.
.
.
Compiling sallyport v0.4.0 (/Users/paul/Documents/GitHub/enarx/enarx/crates/sallyport)
error[E0432]: unresolved import `core::arch::x86_64`
 --> crates/sallyport/src/guest/call/enarxcall/cpuid.rs:9:17
  |
9 | use core::arch::x86_64::CpuidResult;
  |                 ^^^^^^ could not find `x86_64` in `arch`

error[E0432]: unresolved import `core::arch::x86_64`
  --> crates/sallyport/src/guest/handler.rs:24:17
   |
24 | use core::arch::x86_64::CpuidResult;
   |                 ^^^^^^ could not find `x86_64` in `arch`

error[E0432]: unresolved import `core::arch::x86_64`
 --> crates/sallyport/src/host/enarxcall.rs:8:17
  |
8 | use core::arch::x86_64::CpuidResult;
  |                 ^^^^^^ could not find `x86_64` in `arch`

error[E0433]: failed to resolve: could not find `x86_64` in `arch`
   --> crates/sallyport/src/guest/call/syscall/stub.rs:137:41
    |
137 |                 if unsafe { core::arch::x86_64::_rdrand64_step(&mut el) } == 1 {
    |                                         ^^^^^^ could not find `x86_64` in `arch`

error[E0412]: cannot find type `Result` in module `types`
   --> crates/sallyport/src/guest/call/syscall/alloc.rs:114:12
    |
114 |     types::Result<T::Ret>: Into<Result<T::Ret>>,
    |            ^^^^^^ not found in `types`
    |
help: consider importing one of these items
    |
3   | use core::fmt::Result;
    |
3   | use core::result::Result;
    |
3   | use crate::Result;
    |
3   | use crate::guest::enarxcall::types::Result;
    |
      and 2 other candidates

error[E0412]: cannot find type `Result` in module `super`
  --> crates/sallyport/src/guest/call/syscall/types/alloc.rs:39:12
   |
39 |     super::Result<T::Ret>: Into<Result<T::Ret>>,
   |            ^^^^^^ not found in `super`
   |
help: consider importing one of these items
   |
3  | use core::fmt::Result;
   |
3  | use core::result::Result;
   |
3  | use crate::Result;
   |
3  | use crate::guest::enarxcall::types::Result;
   |
     and 2 other candidates

error[E0412]: cannot find type `Result` in module `super`
  --> crates/sallyport/src/guest/call/syscall/types/alloc.rs:47:25
   |
47 |         let res: super::Result<T::Ret> = ret.into();
   |                         ^^^^^^ not found in `super`
   |
help: consider importing one of these items
   |
3  | use core::fmt::Result;
   |
3  | use core::result::Result;
   |
3  | use crate::Result;
   |
3  | use crate::guest::enarxcall::types::Result;
   |
     and 2 other candidates

warning: unused import: `core::arch::asm`
 --> crates/sallyport/src/host/enarxcall.rs:7:5
  |
7 | use core::arch::asm;
  |     ^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

error: invalid register `rax`: unknown register
  --> crates/sallyport/src/host/syscall.rs:36:9
   |
36 |         inlateout("rax") self.num as usize => *self.ret[0],
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rcx`: unknown register
  --> crates/sallyport/src/host/syscall.rs:37:9
   |
37 |         lateout("rcx") _, // clobbered
   |         ^^^^^^^^^^^^^^^^

error: invalid register `r11`: unknown register
  --> crates/sallyport/src/host/syscall.rs:38:9
   |
38 |         lateout("r11") _, // clobbered
   |         ^^^^^^^^^^^^^^^^

error: invalid register `rax`: unknown register
  --> crates/sallyport/src/host/syscall.rs:48:9
   |
48 |         inlateout("rax") self.num as usize => *self.ret[0],
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rdi`: unknown register
  --> crates/sallyport/src/host/syscall.rs:49:9
   |
49 |         in("rdi") self.argv[0],
   |         ^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rcx`: unknown register
  --> crates/sallyport/src/host/syscall.rs:50:9
   |
50 |         lateout("rcx") _, // clobbered
   |         ^^^^^^^^^^^^^^^^

error: invalid register `r11`: unknown register
  --> crates/sallyport/src/host/syscall.rs:51:9
   |
51 |         lateout("r11") _, // clobbered
   |         ^^^^^^^^^^^^^^^^

error: invalid register `rax`: unknown register
  --> crates/sallyport/src/host/syscall.rs:61:9
   |
61 |         inlateout("rax") self.num as usize => *self.ret[0],
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rdi`: unknown register
  --> crates/sallyport/src/host/syscall.rs:62:9
   |
62 |         in("rdi") self.argv[0],
   |         ^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rsi`: unknown register
  --> crates/sallyport/src/host/syscall.rs:63:9
   |
63 |         in("rsi") self.argv[1],
   |         ^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rcx`: unknown register
  --> crates/sallyport/src/host/syscall.rs:64:9
   |
64 |         lateout("rcx") _, // clobbered
   |         ^^^^^^^^^^^^^^^^

error: invalid register `r11`: unknown register
  --> crates/sallyport/src/host/syscall.rs:65:9
   |
65 |         lateout("r11") _, // clobbered
   |         ^^^^^^^^^^^^^^^^

error: invalid register `rax`: unknown register
  --> crates/sallyport/src/host/syscall.rs:75:9
   |
75 |         inlateout("rax") self.num as usize => *self.ret[0],
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rdi`: unknown register
  --> crates/sallyport/src/host/syscall.rs:76:9
   |
76 |         in("rdi") self.argv[0],
   |         ^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rsi`: unknown register
  --> crates/sallyport/src/host/syscall.rs:77:9
   |
77 |         in("rsi") self.argv[1],
   |         ^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rdx`: unknown register
  --> crates/sallyport/src/host/syscall.rs:78:9
   |
78 |         in("rdx") self.argv[2],
   |         ^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rcx`: unknown register
  --> crates/sallyport/src/host/syscall.rs:79:9
   |
79 |         lateout("rcx") _, // clobbered
   |         ^^^^^^^^^^^^^^^^

error: invalid register `r11`: unknown register
  --> crates/sallyport/src/host/syscall.rs:80:9
   |
80 |         lateout("r11") _, // clobbered
   |         ^^^^^^^^^^^^^^^^

error: invalid register `rax`: unknown register
  --> crates/sallyport/src/host/syscall.rs:90:9
   |
90 |         inlateout("rax") self.num as usize => *self.ret[0],
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rdi`: unknown register
  --> crates/sallyport/src/host/syscall.rs:91:9
   |
91 |         in("rdi") self.argv[0],
   |         ^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rsi`: unknown register
  --> crates/sallyport/src/host/syscall.rs:92:9
   |
92 |         in("rsi") self.argv[1],
   |         ^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rdx`: unknown register
  --> crates/sallyport/src/host/syscall.rs:93:9
   |
93 |         in("rdx") self.argv[2],
   |         ^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `r10`: unknown register
  --> crates/sallyport/src/host/syscall.rs:94:9
   |
94 |         in("r10") self.argv[3],
   |         ^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rcx`: unknown register
  --> crates/sallyport/src/host/syscall.rs:95:9
   |
95 |         lateout("rcx") _, // clobbered
   |         ^^^^^^^^^^^^^^^^

error: invalid register `r11`: unknown register
  --> crates/sallyport/src/host/syscall.rs:96:9
   |
96 |         lateout("r11") _, // clobbered
   |         ^^^^^^^^^^^^^^^^

error: invalid register `rax`: unknown register
   --> crates/sallyport/src/host/syscall.rs:106:9
    |
106 |         inlateout("rax") self.num as usize => *self.ret[0],
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rdi`: unknown register
   --> crates/sallyport/src/host/syscall.rs:107:9
    |
107 |         in("rdi") self.argv[0],
    |         ^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rsi`: unknown register
   --> crates/sallyport/src/host/syscall.rs:108:9
    |
108 |         in("rsi") self.argv[1],
    |         ^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rdx`: unknown register
   --> crates/sallyport/src/host/syscall.rs:109:9
    |
109 |         in("rdx") self.argv[2],
    |         ^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `r10`: unknown register
   --> crates/sallyport/src/host/syscall.rs:110:9
    |
110 |         in("r10") self.argv[3],
    |         ^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `r8`: unknown register
   --> crates/sallyport/src/host/syscall.rs:111:9
    |
111 |         in("r8") self.argv[4],
    |         ^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rcx`: unknown register
   --> crates/sallyport/src/host/syscall.rs:112:9
    |
112 |         lateout("rcx") _, // clobbered
    |         ^^^^^^^^^^^^^^^^

error: invalid register `r11`: unknown register
   --> crates/sallyport/src/host/syscall.rs:113:9
    |
113 |         lateout("r11") _, // clobbered
    |         ^^^^^^^^^^^^^^^^

error: invalid register `rax`: unknown register
   --> crates/sallyport/src/host/syscall.rs:123:9
    |
123 |         inlateout("rax") self.num as usize => *self.ret[0],
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rdi`: unknown register
   --> crates/sallyport/src/host/syscall.rs:124:9
    |
124 |         in("rdi") self.argv[0],
    |         ^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rsi`: unknown register
   --> crates/sallyport/src/host/syscall.rs:125:9
    |
125 |         in("rsi") self.argv[1],
    |         ^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rdx`: unknown register
   --> crates/sallyport/src/host/syscall.rs:126:9
    |
126 |         in("rdx") self.argv[2],
    |         ^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `r10`: unknown register
   --> crates/sallyport/src/host/syscall.rs:127:9
    |
127 |         in("r10") self.argv[3],
    |         ^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `r8`: unknown register
   --> crates/sallyport/src/host/syscall.rs:128:9
    |
128 |         in("r8") self.argv[4],
    |         ^^^^^^^^^^^^^^^^^^^^^

error: invalid register `r9`: unknown register
   --> crates/sallyport/src/host/syscall.rs:129:9
    |
129 |         in("r9") self.argv[5],
    |         ^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rcx`: unknown register
   --> crates/sallyport/src/host/syscall.rs:130:9
    |
130 |         lateout("rcx") _, // clobbered
    |         ^^^^^^^^^^^^^^^^

error: invalid register `r11`: unknown register
   --> crates/sallyport/src/host/syscall.rs:131:9
    |
131 |         lateout("r11") _, // clobbered
    |         ^^^^^^^^^^^^^^^^

error: invalid register `rax`: unknown register
   --> crates/sallyport/src/item/enarxcall/sgx.rs:177:17
    |
177 |                 in("rax") EREPORT,
    |                 ^^^^^^^^^^^^^^^^^

error: invalid register `rcx`: unknown register
   --> crates/sallyport/src/item/enarxcall/sgx.rs:178:17
    |
178 |                 in("rcx") reportdata.0.as_ptr(),
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rdx`: unknown register
   --> crates/sallyport/src/item/enarxcall/sgx.rs:179:17
    |
179 |                 in("rdx") report.as_mut_ptr(),
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

   Compiling indexmap v1.8.1
   Compiling cranelift-entity v0.83.0
   Compiling bincode v1.3.3
   Compiling url v2.2.2
   Compiling toml v0.5.9
   Compiling serde_json v1.0.79
error[E0282]: type annotations needed
   --> crates/sallyport/src/guest/call/syscall/alloc.rs:112:10
    |
112 | impl<'a, T: Alloc<'a>> super::super::Alloc<'a, alloc::kind::Syscall> for T
    |          ^ cannot infer type for type parameter `T`

error[E0282]: type annotations needed
   --> crates/sallyport/src/guest/call/syscall/alloc.rs:116:19
    |
116 |     type Staged = StagedAlloc<'a, T>;
    |                   ^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `T`

error[E0282]: type annotations needed
   --> crates/sallyport/src/guest/call/syscall/alloc.rs:117:22
    |
117 |     type Committed = CommittedAlloc<'a, T>;
    |                      ^^^^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `T`

error[E0282]: type annotations needed
   --> crates/sallyport/src/guest/call/syscall/alloc.rs:121:8
    |
121 |     fn stage(self, alloc: &mut impl Allocator) -> Result<Self::Staged> {
    |        ^^^^^ cannot infer type for type parameter `T`

error[E0283]: type annotations needed
  --> crates/sallyport/src/guest/call/syscall/types/alloc.rs:37:24
   |
37 | impl<'a, T: Alloc<'a>> Collect for CommittedAlloc<'a, T>
   |                        ^^^^^^^ cannot infer type for struct `call::syscall::types::alloc::CommittedAlloc<'a, T>`
   |
note: multiple `impl`s satisfying `call::syscall::types::alloc::CommittedAlloc<'a, T>: guest::alloc::Collect` found
  --> crates/sallyport/src/guest/call/syscall/types/alloc.rs:37:1
   |
37 | / impl<'a, T: Alloc<'a>> Collect for CommittedAlloc<'a, T>
38 | | where
39 | |     super::Result<T::Ret>: Into<Result<T::Ret>>,
40 | | {
...  |
49 | |     }
50 | | }
   | |_^
   |
  ::: crates/sallyport/src/guest/call/stub.rs:28:1
   |
28 |   impl<T: Stub> Collect for T {
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0284]: type annotations needed: cannot satisfy `<call::syscall::types::alloc::CommittedAlloc<'a, T> as guest::alloc::Collect>::Item == _`
  --> crates/sallyport/src/guest/call/syscall/types/alloc.rs:44:8
   |
44 |     fn collect(self, col: &impl Collector) -> Self::Item {
   |        ^^^^^^^ cannot satisfy `<call::syscall::types::alloc::CommittedAlloc<'a, T> as guest::alloc::Collect>::Item == _`

Some errors have detailed explanations: E0282, E0283, E0284, E0412, E0432, E0433.
For more information about an error, try `rustc --explain E0282`.
warning: `sallyport` (lib) generated 1 warning
error: could not compile `sallyport` due to 58 previous errors; 1 warning emitted
warning: build failed, waiting for other jobs to finish...
```

**Comment by rjzak** [comment]
I was able to run a console wasm file on my M1 Mac, macOS 12.3.1.

**Comment by rjzak** [comment]
Tests seem to pass, don't know about the error.

`cargo test --no-default-features --features=backend-nil --test wasmldr_tests`

```
running 9 tests
test echo ... ok
test memory_stress_test ... ok
test return_1 ... ok
test wasi_snapshot1 ... ok
Error: default export in '' is not a function
test no_export ... ok
test hello_wasi_snapshot1 ... ok
test check_tcp ... ok
test memspike ... ok
test zerooneone ... ok

test result: ok. 9 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 13.36s
```

**Review by platten — APPROVED** [review]
LGTM!

File: .cargo/config [hunk 1]
```diff
@@ -6,5 +6,10 @@ rustflags = [
     "-C", "linker=gcc",
 ]
 
+[target.x86_64-apple-darwin]
```
Comment (rjzak) [inline .cargo/config:9]: Is this needed? If so, needed for arm64?

**Comment by npmccallum** [comment]
@haraldh Please rebase.

**Comment by rjzak** [comment]
I tested this on the Raspberry Pi, and it didn't compile due to arguments passed to the compiler.

```pi@raspberrypi:~/Documents/enarx-nil $ cargo build --no-default-features --features=backend-nil
   Compiling wasmtime-runtime v0.36.0
   Compiling ring v0.16.20
   Compiling psm v0.1.18
The following warnings were emitted during compilation:

warning: cc1: error: unrecognized command-line option ‘-m64’

error: failed to run custom build command for `wasmtime-runtime v0.36.0`

Caused by:
  process didn't exit successfully: `/home/pi/Documents/enarx-nil/target/debug/build/wasmtime-runtime-512f42b4316e525e/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-changed=src/helpers.c
  TARGET = Some("x86_64-unknown-linux-musl")
  OPT_LEVEL = Some("0")
  HOST = Some("aarch64-unknown-linux-gnu")
  CC_x86_64-unknown-linux-musl = None
  CC_x86_64_unknown_linux_musl = None
  TARGET_CC = None
  CC = None
  CROSS_COMPILE = None
  CFLAGS_x86_64-unknown-linux-musl = None
  CFLAGS_x86_64_unknown_linux_musl = None
  TARGET_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("true")
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,llvm14-builtins-abi,sse,sse2")
  running: "musl-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-Wall" "-Wextra" "-DCFG_TARGET_OS_linux" "-o" "/home/pi/Documents/enarx-nil/target/x86_64-unknown-linux-musl/debug/build/wasmtime-runtime-d188d93e978ac2e8/out/src/helpers.o" "-c" "src/helpers.c"
  cargo:warning=cc1: error: unrecognized command-line option ‘-m64’
  exit status: 1

  --- stderr


  error occurred: Command "musl-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-Wall" "-Wextra" "-DCFG_TARGET_OS_linux" "-o" "/home/pi/Documents/enarx-nil/target/x86_64-unknown-linux-musl/debug/build/wasmtime-runtime-d188d93e978ac2e8/out/src/helpers.o" "-c" "src/helpers.c" with args "musl-gcc" did not execute successfully (status code exit status: 1).


warning: build failed, waiting for other jobs to finish...
The following warnings were emitted during compilation:

warning: cc1: error: unrecognized command-line option ‘-m64’

error: failed to run custom build command for `psm v0.1.18`

Caused by:
  process didn't exit successfully: `/home/pi/Documents/enarx-nil/target/debug/build/psm-b4980df40024a355/build-script-build` (exit status: 1)
  --- stdout
  OPT_LEVEL = Some("0")
  TARGET = Some("x86_64-unknown-linux-musl")
  HOST = Some("aarch64-unknown-linux-gnu")
  CC_x86_64-unknown-linux-musl = None
  CC_x86_64_unknown_linux_musl = None
  TARGET_CC = None
  CC = None
  CROSS_COMPILE = None
  CFLAGS_x86_64-unknown-linux-musl = None
  CFLAGS_x86_64_unknown_linux_musl = None
  TARGET_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("true")
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,llvm14-builtins-abi,sse,sse2")
  cargo:rustc-cfg=asm
  cargo:rustc-cfg=switchable_stack
  running: "musl-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-Wall" "-Wextra" "-xassembler-with-cpp" "-DCFG_TARGET_OS_linux" "-DCFG_TARGET_ARCH_x86_64" "-DCFG_TARGET_ENV_musl" "-o" "/home/pi/Documents/enarx-nil/target/x86_64-unknown-linux-musl/debug/build/psm-9feac2499397127c/out/src/arch/x86_64.o" "-c" "src/arch/x86_64.s"
  cargo:warning=cc1: error: unrecognized command-line option ‘-m64’
  exit status: 1

  --- stderr


  error occurred: Command "musl-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-Wall" "-Wextra" "-xassembler-with-cpp" "-DCFG_TARGET_OS_linux" "-DCFG_TARGET_ARCH_x86_64" "-DCFG_TARGET_ENV_musl" "-o" "/home/pi/Documents/enarx-nil/target/x86_64-unknown-linux-musl/debug/build/psm-9feac2499397127c/out/src/arch/x86_64.o" "-c" "src/arch/x86_64.s" with args "musl-gcc" did not execute successfully (status code exit status: 1).


error: failed to run custom build command for `ring v0.16.20`

Caused by:
  process didn't exit successfully: `/home/pi/Documents/enarx-nil/target/debug/build/ring-06a53c715f5675c3/build-script-build` (exit status: 101)
  --- stdout
  OPT_LEVEL = Some("0")
  TARGET = Some("x86_64-unknown-linux-musl")
  HOST = Some("aarch64-unknown-linux-gnu")
  CC_x86_64-unknown-linux-musl = None
  CC_x86_64_unknown_linux_musl = None
  TARGET_CC = None
  CC = None
  CROSS_COMPILE = None
  CFLAGS_x86_64-unknown-linux-musl = None
  CFLAGS_x86_64_unknown_linux_musl = None
  TARGET_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("true")
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,llvm14-builtins-abi,sse,sse2")

  --- stderr
  running "musl-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-I" "include" "-Wall" "-Wextra" "-pedantic" "-pedantic-errors" "-Wall" "-Wextra" "-Wcast-align" "-Wcast-qual" "-Wconversion" "-Wenum-compare" "-Wfloat-equal" "-Wformat=2" "-Winline" "-Winvalid-pch" "-Wmissing-field-initializers" "-Wmissing-include-dirs" "-Wredundant-decls" "-Wshadow" "-Wsign-compare" "-Wsign-conversion" "-Wundef" "-Wuninitialized" "-Wwrite-strings" "-fno-strict-aliasing" "-fvisibility=hidden" "-fstack-protector" "-g3" "-U_FORTIFY_SOURCE" "-DNDEBUG" "-c" "-o/home/pi/Documents/enarx-nil/target/x86_64-unknown-linux-musl/debug/build/ring-a4597aa78190d992/out/aesni-x86_64-elf.o" "/home/pi/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.16.20/pregenerated/aesni-x86_64-elf.S"
  cc1: error: unrecognized command-line option ‘-m64’
  thread 'main' panicked at 'execution failed', /home/pi/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.16.20/build.rs:656:9
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

File: .cargo/config (hunk 1 above)
Comment (haraldh) [reply .cargo/config]: removed

**Comment by rjzak** [comment]
> yeah `-m64` ... your PI might be 32 bit only :)

It's Raspberry Pi 4 with 64-bit OS.

```
pi@raspberrypi:~/Documents/enarx-nil $ uname -a
Linux raspberrypi 5.15.32-v8+ #1538 SMP PREEMPT Thu Mar 31 19:40:39 BST 2022 aarch64 GNU/Linux
```

File: Cargo.toml [hunk 2]
```diff
@@ -61,6 +61,7 @@ vdso = { version = "0.2", optional = true, default-features = false }
 x86_64 = { version = "0.14.9", optional = true, default-features = false }
 
 # binary dependencies
+[target.'cfg(target_os = "linux")'.dependencies]
 # FIXME: The enarx-exec-wasmtime-bin crate should be merged with enarx-exec-wasmtime again, see https://github.com/enarx/enarx/issues/1798
 enarx-exec-wasmtime-bin = { version = "0.4.0", path = "crates/exec-wasmtime-bin", artifact = "bin", target = "x86_64-unknown-linux-musl", default-features = false }
```
Comment (rjzak) [inline Cargo.toml:66]: Is it possible to express the targets as linux but without the arch? Thinking about Rasp Pi.

**Comment by rjzak** [comment]
Still pulls x86_64-unknown-linux-gnu target

```
pi@raspberrypi:~/Documents/enarx-nil $ cargo build
info: syncing channel updates for 'nightly-2022-04-11-aarch64-unknown-linux-gnu'
info: latest update on 2022-04-11, rust version 1.62.0-nightly (1f7fb6413 2022-04-10)
info: component 'clippy' for target 'aarch64-unknown-linux-gnu' is up to date
info: component 'miri' for target 'aarch64-unknown-linux-gnu' is up to date
info: component 'rust-src' is up to date
info: component 'rust-std' for target 'wasm32-wasi' is up to date
info: component 'rustfmt' for target 'aarch64-unknown-linux-gnu' is up to date
info: downloading component 'rust-std' for 'x86_64-unknown-linux-gnu'
^C
pi@raspberrypi:~/Documents/enarx-nil $ git log -1
commit e2483a7f38910e3b1b9950ec1e3efbce250ff081 (HEAD -> macos_nil_support)
Author: Harald Hoyer <harald@profian.com>
Date:   Fri Apr 29 15:50:18 2022 +0200

    ci: add MacOS test
    
    Signed-off-by: Harald Hoyer <harald@profian.com>
pi@raspberrypi:~/Documents/enarx-nil $ uname -a
Linux raspberrypi 5.15.32-v8+ #1538 SMP PREEMPT Thu Mar 31 19:40:39 BST 2022 aarch64 GNU/Linux
```

**Comment by haraldh** [comment]
> Still pulls x86_64-unknown-linux-gnu target

@rjzak try to remove them from `rust-toolchain.toml`

**Comment by rjzak** [comment]
Building on RaspPi4 fails at the last minute due to Intel assembly

```
   Compiling enarx-exec-wasmtime v0.4.0 (/home/pi/Documents/enarx-nil/crates/exec-wasmtime)
error: invalid register `rax`: unknown register
  --> crates/exec-wasmtime/src/loader/configured/platform.rs:55:17
   |
55 |                 lateout("rax") rax,
   |                 ^^^^^^^^^^^^^^^^^^

error: invalid register `rdx`: unknown register
  --> crates/exec-wasmtime/src/loader/configured/platform.rs:56:17
   |
56 |                 lateout("rdx") rdx,
   |                 ^^^^^^^^^^^^^^^^^^

error: invalid register `rax`: unknown register
  --> crates/exec-wasmtime/src/loader/configured/platform.rs:57:17
   |
57 |                 in("rax") Self::SYS_GETATT,
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rdi`: unknown register
  --> crates/exec-wasmtime/src/loader/configured/platform.rs:58:17
   |
58 |                 in("rdi") nonce.map(|x| x.as_ptr()).unwrap_or_else(null),
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rsi`: unknown register
  --> crates/exec-wasmtime/src/loader/configured/platform.rs:59:17
   |
59 |                 in("rsi") nonce.map(|x| x.len()).unwrap_or_default(),
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rdx`: unknown register
  --> crates/exec-wasmtime/src/loader/configured/platform.rs:60:17
   |
60 |                 in("rdx") buf.as_mut().map(|x| x.as_mut_ptr()).unwrap_or_else(null_mut),
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `r10`: unknown register
  --> crates/exec-wasmtime/src/loader/configured/platform.rs:61:17
   |
61 |                 in("r10") buf.map(|x| x.len()).unwrap_or_default(),
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: invalid register `rcx`: unknown register
  --> crates/exec-wasmtime/src/loader/configured/platform.rs:62:17
   |
62 |                 lateout("rcx") _, // clobbered
   |                 ^^^^^^^^^^^^^^^^

error: invalid register `r11`: unknown register
  --> crates/exec-wasmtime/src/loader/configured/platform.rs:63:17
   |
63 |                 lateout("r11") _, // clobbered
   |                 ^^^^^^^^^^^^^^^^

error: could not compile `enarx-exec-wasmtime` due to 9 previous errors

real	11m40.614s
user	31m2.066s
sys	1m20.375s
```

**Comment by rjzak** [comment]
I was able to run a Wasm file with Enarx on my Raspberry Pi 4.

CC: @MikeCamel

**Comment by npmccallum** [comment]
@haraldh I'm still unable to get this to work on my M1 Mac.

```
cargo build --no-default-features --features backend-nil
   Compiling hashbrown v0.11.2
   Compiling bitflags v1.3.2
   Compiling itoa v1.0.1
   Compiling fallible-iterator v0.2.0
   Compiling stable_deref_trait v1.2.0
   Compiling smallvec v1.8.0
   Compiling wasmparser v0.83.0
   Compiling lazy_static v1.4.0
   Compiling rustc-hash v1.1.0
   Compiling ppv-lite86 v0.2.16
   Compiling ambient-authority v0.0.1
   Compiling more-asserts v0.2.2
   Compiling either v1.6.1
   Compiling ipnet v2.5.0
   Compiling maybe-owned v0.3.4
   Compiling untrusted v0.7.1
   Compiling rustc-demangle v0.1.21
   Compiling once_cell v1.10.0
   Compiling const-oid v0.9.0
   Compiling flagset v0.4.3
   Compiling tinyvec_macros v0.1.0
   Compiling pin-project-lite v0.2.8
   Compiling matches v0.1.9
   Compiling unicode-bidi v0.3.7
   Compiling percent-encoding v2.1.0
   Compiling base64ct v1.5.0
   Compiling os_str_bytes v6.0.0
   Compiling base64 v0.13.0
   Compiling chunked_transfer v1.4.0
   Compiling textwrap v0.15.0
   Compiling ryu v1.0.9
   Compiling zeroize v1.5.4
   Compiling colorful v0.2.1
   Compiling proc-macro2 v1.0.37
error: failed to run custom build command for `proc-macro2 v1.0.37`

Caused by:
  process didn't exit successfully: `/Users/npmccallum/enarx/target/debug/build/proc-macro2-1d6818e6a9f93074/build-script-build` (signal: 9, SIGKILL: kill)
warning: build failed, waiting for other jobs to finish...
```

It appears to successfully build that build script. But then we get `SIGKILL` when the build script tries to run. It seems to be the right kind of file.

```
file /Users/npmccallum/enarx/target/debug/build/proc-macro2-1d6818e6a9f93074/build-script-build
/Users/npmccallum/enarx/target/debug/build/proc-macro2-1d6818e6a9f93074/build-script-build: Mach-O executable arm64
```

However...

```
lldb /Users/npmccallum/enarx/target/debug/build/proc-macro2-1d6818e6a9f93074/build-script-build
(lldb) target create "/Users/npmccallum/enarx/target/debug/build/proc-macro2-1d6818e6a9f93074/build-script-build"
warning: (arm64) /Users/npmccallum/enarx/target/debug/build/proc-macro2-1d6818e6a9f93074/build-script-build load command 3 LC_SEGMENT has a fileoff + filesize (0x50000) that extends beyond the end of the file (0x4f300), the segment will be truncated to match
Current executable set to '/Users/npmccallum/enarx/target/debug/build/proc-macro2-1d6818e6a9f93074/build-script-build' (arm64).
(lldb) run
error: Malformed Mach-o file
```

**Comment by haraldh** [comment]
> Building on RaspPi4 fails at the last minute due to Intel assembly

This one was fixed.

**Review by jarkkojs — APPROVED** [review]

**Comment by bstrie** [comment]
LGTM, but holding off on approving until we can confirm that it works on M1.

**Comment by platten** [comment]
> LGTM, but holding off on approving until we can confirm that it works on M1.

Works on M1.

**Review by npmccallum — APPROVED** [review]

**Pull request merged**